File: microservices/story-service/src/main/java/com/storyapp/story/service/TrainingHistoryService.java
package com.storyapp.story.service;

import com.storyapp.story.dto.TrainingHistoryRequest;
import com.storyapp.story.model.TrainingHistory;
import com.storyapp.story.repository.TrainingHistoryRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
@Slf4j
public class TrainingHistoryService {
    
    private final TrainingHistoryRepository trainingHistoryRepository;
    private static final DateTimeFormatter ISO_FORMATTER = DateTimeFormatter.ISO_DATE_TIME;
    
    /**
     * Save a new training history record
     */
    @Transactional
    public TrainingHistory saveTrainingHistory(TrainingHistoryRequest request) {
        log.info("Saving training history for user {} with session ID {}", 
                 request.getUserId(), request.getSessionId());
        
        TrainingHistory history = TrainingHistory.builder()
                .userId(request.getUserId())
                .sessionId(request.getSessionId())
                .numExamplesRequested(request.getNumExamplesRequested())
                .numExamplesSuccessful(request.getNumExamplesSuccessful())
                .numExamplesFailed(request.getNumExamplesFailed())
                .genresSelected(request.getGenresSelected())
                .storedInMemory(request.getStoredInMemory())
                .savedToDatabase(request.getSavedToDatabase())
                .storiesSavedCount(request.getStoriesSavedCount())
                .averageQuality(request.getAverageQuality())
                .minQuality(request.getMinQuality())
                .maxQuality(request.getMaxQuality())
                .bestPerformingModel(request.getBestPerformingModel())
                .modelPerformanceJson(request.getModelPerformanceJson())
                .genreDistributionJson(request.getGenreDistributionJson())
                .totalTimeSeconds(request.getTotalTimeSeconds())
                .trainingExamplesBefore(request.getTrainingExamplesBefore())
                .trainingExamplesAfter(request.getTrainingExamplesAfter())
                .finetuningTriggered(request.getFinetuningTriggered())
                .startedAt(parseDateTime(request.getStartedAt()))
                .completedAt(parseDateTime(request.getCompletedAt()))
                .build();
        
        TrainingHistory saved = trainingHistoryRepository.save(history);
        log.info("Training history saved with ID: {}", saved.getId());
        
        return saved;
    }
    
    /**
     * Get all training history for a user
     */
    public List<TrainingHistory> getUserTrainingHistory(Long userId) {
        log.info("Fetching training history for user {}", userId);
        return trainingHistoryRepository.findByUserIdOrderByStartedAtDesc(userId);
    }
    
    /**
     * Get paginated training history for a user
     */
    public Page<TrainingHistory> getUserTrainingHistoryPaginated(Long userId, int page, int size) {
        log.info("Fetching paginated training history for user {} (page: {}, size: {})", 
                 userId, page, size);
        Pageable pageable = PageRequest.of(page, size);
        return trainingHistoryRepository.findByUserIdOrderByStartedAtDesc(userId, pageable);
    }
    
    /**
     * Get a specific training session by session ID
     */
    public Optional<TrainingHistory> getTrainingSessionBySessionId(String sessionId) {
        log.info("Fetching training session with session ID: {}", sessionId);
        return trainingHistoryRepository.findBySessionId(sessionId);
    }
    
    /**
     * Get training history within a date range
     */
    public List<TrainingHistory> getUserTrainingHistoryByDateRange(
            Long userId, LocalDateTime startDate, LocalDateTime endDate) {
        log.info("Fetching training history for user {} between {} and {}", 
                 userId, startDate, endDate);
        return trainingHistoryRepository.findByUserIdAndDateRange(userId, startDate, endDate);
    }
    
    /**
     * Get count of training sessions for a user
     */
    public long getUserTrainingSessionCount(Long userId) {
        return trainingHistoryRepository.countByUserId(userId);
    }
    
    /**
     * Get sessions that triggered fine-tuning
     */
    public List<TrainingHistory> getFineTuningTriggeringSessions(Long userId) {
        log.info("Fetching fine-tuning triggering sessions for user {}", userId);
        return trainingHistoryRepository.findByUserIdAndFinetuningTriggeredTrue(userId);
    }
    
    /**
     * Get recent training sessions
     */
    public Page<TrainingHistory> getRecentTrainingSessions(Long userId, int limit) {
        log.info("Fetching {} most recent training sessions for user {}", limit, userId);
        Pageable pageable = PageRequest.of(0, limit);
        return trainingHistoryRepository.findByUserIdOrderByCompletedAtDesc(userId, pageable);
    }
    
    /**
     * Parse ISO datetime string to LocalDateTime
     */
    private LocalDateTime parseDateTime(String isoDateString) {
        try {
            return LocalDateTime.parse(isoDateString, ISO_FORMATTER);
        } catch (Exception e) {
            log.warn("Failed to parse datetime: {}, using current time", isoDateString);
            return LocalDateTime.now();
        }
    }
}

File: Frontend/src/api/trainingHistory.service.ts
import httpClient from './httpClient.js';

const STORY_SERVICE_URL = 'http://localhost:8082/api';

export interface TrainingHistory {
  id: number;
  userId: number;
  sessionId: string;
  numExamplesRequested: number;
  numExamplesSuccessful: number;
  numExamplesFailed: number;
  genresSelected: string;
  storedInMemory: boolean;
  savedToDatabase: boolean;
  storiesSavedCount: number;
  averageQuality: number;
  minQuality: number;
  maxQuality: number;
  bestPerformingModel: string;
  modelPerformanceJson: string;
  genreDistributionJson: string;
  totalTimeSeconds: number;
  trainingExamplesBefore: number;
  trainingExamplesAfter: number;
  finetuningTriggered: boolean;
  startedAt: string;
  completedAt: string;
  createdAt: string;
}

export interface TrainingStats {
  totalSessions: number;
  totalExamplesGenerated: number;
  totalStoriesSaved: number;
  averageQuality: number;
  totalTrainingTimeSeconds: number;
  fineTuningSessionsCount: number;
}

/**
 * Get all training history for a user
 */
export const getUserTrainingHistory = async (userId: number): Promise<TrainingHistory[]> => {
  try {
    const response = await fetch(`${STORY_SERVICE_URL}/training-history/user/${userId}`, {
      headers: { 'Content-Type': 'application/json' }
    });
    if (!response.ok) return [];
    return await response.json();
  } catch {
    return [];
  }
};

/**
 * Get paginated training history
 */
export const getUserTrainingHistoryPaginated = async (
  userId: number,
  page: number = 0,
  size: number = 10
): Promise<{ content: TrainingHistory[]; totalPages: number; totalElements: number }> => {
  const response = await httpClient.get(
    `${STORY_SERVICE_URL}/training-history/user/${userId}/paginated`,
    { params: { page, size } }
  );
  return response.data;
};

/**
 * Get a specific training session by session ID
 */
export const getTrainingSessionBySessionId = async (sessionId: string): Promise<TrainingHistory> => {
  const response = await httpClient.get(`${STORY_SERVICE_URL}/training-history/session/${sessionId}`);
  return response.data;
};

/**
 * Get training statistics for a user
 */
export const getUserTrainingStats = async (userId: number): Promise<TrainingStats> => {
  try {
    const response = await fetch(`${STORY_SERVICE_URL}/training-history/user/${userId}/stats`, {
      headers: { 'Content-Type': 'application/json' }
    });
    if (!response.ok) {
      return {
        totalSessions: 0,
        totalExamplesGenerated: 0,
        totalStoriesSaved: 0,
        averageQuality: 0,
        totalTrainingTimeSeconds: 0,
        fineTuningSessionsCount: 0
      };
    }
    return await response.json();
  } catch {
    return {
      totalSessions: 0,
      totalExamplesGenerated: 0,
      totalStoriesSaved: 0,
      averageQuality: 0,
      totalTrainingTimeSeconds: 0,
      fineTuningSessionsCount: 0
    };
  }
};

/**
 * Get recent training sessions
 */
export const getRecentTrainingSessions = async (
  userId: number,
  limit: number = 5
): Promise<{ content: TrainingHistory[] }> => {
  const response = await httpClient.get(
    `${STORY_SERVICE_URL}/training-history/user/${userId}/recent`,
    { params: { limit } }
  );
  return response.data;
};
